Firestore/core/test/unit/testutil/testutil.h

```diff
@@ -179,24 +191,62 @@ google_firestore_v1_Value AddPairs(const google_firestore_v1_Value& prior,
  *     be passed to Value().
  */
 template <typename... Args>
-google_firestore_v1_Value MakeMap(Args... key_value_pairs) {
-  google_firestore_v1_Value map_value{};
-  map_value.which_value_type = google_firestore_v1_Value_map_value_tag;
-  return AddPairs(map_value, key_value_pairs...);
+nanopb::Message<google_firestore_v1_Value> MakeMap(Args... key_value_pairs) {
+  nanopb::Message<google_firestore_v1_Value> map_value{{}};
```

Comment:
Ok, I'm not 100% sure, but I think I understand the general idea. Reading from a union member that wasn't written before is undefined behavior. It's not enough to set the tag to the correct value, you also need to initialize the union member (`map_value->map_value = {}`). Removing the braces leaves the union in a state where only the first member is initialized, so when subsequent code tries to access `map_value->map_value`, it reads random data.

What I'm not sure about is why adding the braces evidently zeroes out the union (that is, it doesn't just set its first member to zero but the whole union to zero) while initializing the proto member in `Message` using `{}` braces only initializes the first union member. The simplest solution seems to be to explicitly initialize the desired union member in this case. Sebastian, how many other cases are there that rely on this zeroing out behavior? I can spend more time digging into this if there are quite a few; if there aren't many, it's probably easier to just explicitly initialize the desired union member.

Reply:
The specific failure I am seeing was that an integration test was using a random value for FieldValue.increment() when the base value did not exist. Since I have seen a bunch of these failures during my earlier testing (before migrating to Message), I added explicit initialization to all Value protos. The failures were pretty random and hard to debug. It felt safe to uniformly initialize nanopb messages rather than cherrypick the ones that caused known problems. 

The specific issue in this PR was https://github.com/firebase/firebase-ios-sdk/blob/5b965d421875eb31acc88ebb80c5bf30a7eb8ebc/Firestore/core/src/model/transform_operation.cc#L420, but I am not sure if this is the only issue since the issues are seemingly random.

Reply:
So there are two things:
- I'm not sure why `google_firestore_v1_Value{}` results in the union being zeroed out while `Message<google_firestore_v1_Value>{}` doesn't, even though `Message` initializes `T` using list initialization as well. From what I gather, `google_firestore_v1_Value{}` should invoke aggregate initialization, though I'm not entirely sure whether using aggregate initialization on a union aggregate initializes each member, or only the first member. As a consequence, I'm not 100% sure we can rely on this form of initialization.
- even if we can rely on this, I think some of the current accesses still invoke undefined behavior. Accessing an inactive member of the union is UB, and only one member can be active at any given time. Thus, even if a union has been entirely zeroed out, accessing an element without writing to it first is still UB.

From the second point, I think we should make sure that we always write to a member of the union, making it active, before accessing it, and stop relying on zeroing out the union (FWIW, I think it would be fine in C). Two heuristics to find those places:
- `grep "which_value_type = "` (because we always set the value type before reading from the union);
- `grep "Message<.*\{\}"`.

This definitely includes `zero_value` you mentioned. Running the search, it seems there are quite a few those, but mostly confined to a pretty small set of files.

Feel free to ping me if you'd like to discuss this further.

Reply:
`Message<google_firestore_v1_Value>{}` also zeroes out the union. Dropping the curly braces might not.

I went through the code and added explicit initialization everywhere. I also removed the curly braces where possible. Note that `SharedMessage<goog_firestore_v1_ArrayValue>` still seems to require `{{}}`.

Reply:
The point is, AFAIU, you shouldn't _need_ to zero out the union. If you're writing to the union member that you will be reading later on, it doesn't matter whether it was zeroed out or not. If you're reading a union member and not writing to it first, relying instead on the fact that it was zeroed out, that is undefined behavior. In other words, unions behave differently from regular variables here:
```cpp
int x{};
int y = x; // Ok -- x is zero initialized
union { int a; int b; } u{};
int c = u.b; // UB even if `b` is zeroed out, because `b` is not the active member of the union
```

I _might_ be wrong on this, but that's my reading. The Standard says that only one union member can be active at any given time. To be able to read an arbitrary union member from a zeroed-out union, all of its members would have to be active at the same time, contradicting the previous point. I couldn't find an exception to this rule.

> Note that `SharedMessage<goog_firestore_v1_ArrayValue>` still seems to require `{{}}`.

Yeah, that's because it doesn't have a default constructor. I think we previously replaced some of those using `MakeSharedMessage`, but it's totally optional.